### DISCUSSION TITLE
Matching globs in Lua filter

### DISCUSSION THREAD

**Discussion by andrewpbray** [quarto-dev/quarto-cli#6185] [Q&A] [answered]
I'm working on a Lua filter that will be reading in .qmd files and processing the contents. I would like the user to be able to specify which files to read by passing glob patterns through the document yaml. This is to mirror the syntax used to specify of the contents of a listing: `listing: contents: ...`.

What's the recommended way to expand globs into a list of filenames given the distribution of Lua that comes with Quarto? The general topic is addressed here:

<http://lua-users.org/wiki/FileGlob>

The pure Lua seems like the best bet to me, but I was curious about:

1. What is the best way to embed the functionality of [globtopattern.lua](https://github.com/davidm/lua-glob-pattern/blob/master/lua/globtopattern.lua) into my Lua filter? Convert the module definition to a plain function and include it at the top of my filter?
2. I'm not familiar with the specs on glob patterns and how standardized they are. Should I be worried that the parsing rules in this script might not line up with the specs used elsewhere in Quarto?

**Comment by mcanouil** [comment]
I am not sure what you mean (or how is it relevant) by "given the distribution of Lua that comes with Quarto". Or what you mean by "pure LUA".

LUA filter reads the metadata and the document as the Pandoc AST (see `native` format).
While the metadata is YAML, it is not really standardised, but the AST is the specification used by Pandoc and extended a bit in Quarto to standardised how a document is represented (metadata and content).
So LUA filter do not process qmd files, but AST representation of the files which is different from what you want to do.
Maybe you want to use project scripts such as pre-render scripts, see <https://quarto.org/docs/projects/scripts.html>.

See:
- <https://quarto.org/docs/extensions/lua.html>
- <https://quarto.org/docs/extensions/lua-api.html>
- <https://quarto.org/docs/prerelease/1.4/lua_changes.html#relative-paths-in-require-calls>

**Comment by cscheid** [comment, marked as answer]
> What's the recommended way to expand globs into a list of filenames given the distribution of Lua that comes with Quarto?

Unfortunately, in quarto, glob patterns are resolved in TypeScript, and not in Lua. So you will both need to roll your own, and be on the lookout for mismatched specifications. With that said, their implementation is [not _too_ hairy](https://github.com/denoland/deno_std/blob/main/fs/expand_glob.ts) in case you're feeling like attempting to go for it...

**Reply by andrewpbray** [reply]
Ah got it. I was hunting around for that in the source, and when I couldn't find it I figured it might be done in Typescript.

Wish me luck!
